Firestore/core/src/firebase/firestore/local/local_store.mm

```diff
@@ -0,0 +1,486 @@
+
+      absl::optional<ObjectValue> base_value =
+          mutation.ExtractBaseValue(base_document);
+      if (base_value) {
+        // NOTE: The base state should only be applied if there's some existing
+        // document to override, so use a Precondition of exists=true
+        base_mutations.push_back(PatchMutation(mutation.key(), *base_value,
+                                               base_value->ToFieldMask(),
+                                               Precondition::Exists(true)));
+      }
+    }
+
+    MutationBatch batch = mutation_queue_->AddMutationBatch(
+        local_write_time, std::move(base_mutations), std::move(mutations));
+    MaybeDocumentMap changedDocuments =
+        batch.ApplyToLocalDocumentSet(existing_documents);
+    return LocalWriteResult{batch.batch_id(), std::move(changedDocuments)};
+  });
+}
+
+MaybeDocumentMap LocalStore::AcknowledgeBatch(
+    const MutationBatchResult& batch_result) {
+  return persistence_->Run("Acknowledge batch", [&] {
+    const MutationBatch& batch = batch_result.batch();
+    mutation_queue_->AcknowledgeBatch(batch, batch_result.stream_token());
+    ApplyBatchResult(batch_result);
+    mutation_queue_->PerformConsistencyCheck();
+
+    return local_documents_->GetDocuments(batch.keys());
+  });
+}
+
+void LocalStore::ApplyBatchResult(const MutationBatchResult& batch_result) {
+  const MutationBatch& batch = batch_result.batch();
+  DocumentKeySet docKeys = batch.keys();
+  const DocumentVersionMap& versions = batch_result.doc_versions();
+  for (const DocumentKey& docKey : docKeys) {
+    absl::optional<MaybeDocument> remoteDoc =
+        remote_document_cache_->Get(docKey);
+    absl::optional<MaybeDocument> doc = remoteDoc;
+
+    auto ackVersionIter = versions.find(docKey);
+    HARD_ASSERT(ackVersionIter != versions.end(),
+                "docVersions should contain every doc in the write.");
+    const SnapshotVersion& ackVersion = ackVersionIter->second;
+    if (!doc || doc->version() < ackVersion) {
+      doc = batch.ApplyToRemoteDocument(doc, docKey, batch_result);
+      if (!doc) {
+        HARD_ASSERT(
+            !remoteDoc,
+            "Mutation batch %s applied to document %s resulted in nullopt.",
+            batch.ToString(), util::ToString(remoteDoc));
+      } else {
+        remote_document_cache_->Add(*doc);
+      }
+    }
+  }
+
+  mutation_queue_->RemoveMutationBatch(batch);
+}
+
+MaybeDocumentMap LocalStore::RejectBatch(BatchId batch_id) {
+  return persistence_->Run("Reject batch", [&] {
+    absl::optional<MutationBatch> to_reject =
+        mutation_queue_->LookupMutationBatch(batch_id);
+    HARD_ASSERT(to_reject.has_value(), "Attempt to reject nonexistent batch!");
+
+    mutation_queue_->RemoveMutationBatch(*to_reject);
+    mutation_queue_->PerformConsistencyCheck();
+
+    return local_documents_->GetDocuments(to_reject->keys());
+  });
+}
+
+ByteString LocalStore::GetLastStreamToken() {
```

Comment:
Can this function be `const`? If `GetLastStreamToken` on `mutation_queue_` is `const`, then this one should be as well (otherwise, just ignore).

Reply:
It is not `const` on mutation queue, however it can/should be const at a quick glance.

This should probably be a separate PR though.

Reply:
Sounds good. Please double-check with Gil on whether the mutation queue method should be `const`.